Explorer: support `.base` to initialize parent class from struct

Pixep opened the pull request:
Relates to https://github.com/carbon-language/carbon-lang/issues/1881

- Add support for `.base` field in structs for [parent class initialization](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#constructors)
    - Disabling base class initialization without `.base`
- Support class constructors (`Create() -> Self`) for base classes
- Direct access to base class(es) attributes with `object.var` remains unaffected

Changes:
- Add `TypeChecker::FieldTypesWithBase` to help assessing if a struct with `base` fields can be converted to a class
- Add a new `base_type()` attribute+getter to `NominalClassDeclaration` to as a first step to allow resolving parametrized classes
- Add a new `base` attribute+getter to `NominalClassValue` that contains the base class `NominalClassValue`. It is currently used mainly to get and set members of a class object.
- Add `Interpreter::ConvertClassWithBase` to build `NominalClassValue` from a init struct, that contains `.base` fields with either `NominalClassValue` or `StructValue`
- Add `FindClassField` to find a field in a class or its base classes
- Remove superfluous `ClassDeclaration::base()` in favor of `ClassDeclaration::base_type()`

Limitations;
- Though some work is done in that direction, parametrized base class where time is not know at the declaration site are not supported. Namely the example below does not compile
```
base class A(T:! Type) {}
class B(T:! Type) extends A(T) {}
```
But this one is functional already
```
base class A(T:! Type) {}
class B extends A(i32) {}
```

Pixep committed:
Explorer: support `.base` to initialize parent class from struct

- Add support for `.base` for parent class initialization
- Prevent parent class field initialization without using `.base`

Known limitations:
- Initialization of `base` field using a function or base constructor
  constructor is not supported

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
testdata: add note for expected fail test with constructors

zygoloid commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -427,40 +428,66 @@ auto TypeChecker::FieldTypesImplicitlyConvertible(
+  if (class_type.base().has_value()) {
+    auto base_fields = FieldTypesWithParents(*class_type.base().value());
+    fields.emplace_back(NamedValue{
+        .name = "base", .value = (new StructType(std::move(base_fields)))});
```
I think the type we want for `.base` is `partial BaseClass` (see https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#partial-facet). I think we don't support `partial Class` yet, but using just `Class` seems like a better approximation than recursively converting the base class into a struct.

In particular, conversion from struct to class is one-way, and things like `var d: DerivedClass = {.base = BaseClass.Make(), .derived_member = 5};` are intended to work, but that won't work if it requires a conversion from `BaseClass` to the corresponding struct type to exist.

Oh, but it looks like this is unused right now? If you change it as I suggested above, then I think `StructImplicitlyConvertibleToClass` can be replaced with `return FieldTypesImplicitlyConvertible(source_struct.fields(), FieldTypesWithParents(dest_class), impl_scope);`, which seems like a nice simplification. (It'd probably also make sense to rename to `FieldTypesWithBase` if you make that change.)

Pixep replied:
Agree with the idea of changing the representation. I don't believe we can change from Struct to Class before `TypeChecker::ImplicitlyConvert`, and simplify checks for implicit convertions because this would require knowledge about the destination class to be able to infer the class type of `.base`.

The rest shouldn't a problem!

zygoloid reviewed (commented):
The flattening approach you're using here seems like it's going to result in some complexity: we'll need some way to distinguish between fields of base / derived class with the same name, and handling conversions from struct values that aren't struct literals to class type seems like it'll be challenging.

I wonder if it'd be simpler to change the representation we use for classes with inheritance: instead of flattening into a single struct value, we could have a different struct for each layer of inheritance, so that the value representation used by explorer for `{.base = {.a = 1, .b = 2}, .c = 3} as DerivedClass` would actually be `{.base = {.a = 1, .b = 2}, .c = 3}`. We'll then need to transform conversions from `DerivedClass` to `BaseClass` into accesses of the `.base` member, perhaps in the interpreter's `Convert` function, for both `LValue`s referring to classes and to `NominalClassValue`s.

Pixep committed:
Explorer: separate NominalClassValue and base data

Add a new `base()` method to NominalClassValue to get its base class
object, including its fields.

Pixep committed:
Explorer: Remove dead code

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
Explorer: rename 'FieldTypesWithParents'

... to 'FieldTypesWithBase', as suggested by @zygoloid.

Pixep commented:
> The flattening approach you're using here seems like it's going to result in some complexity: we'll need some way to distinguish between fields of base / derived class with the same name, and handling conversions from struct values that aren't struct literals to class type seems like it'll be challenging.
> 
> I wonder if it'd be simpler to change the representation we use for classes with inheritance: instead of flattening into a single struct value, we could have a different struct for each layer of inheritance, so that the value representation used by explorer for `{.base = {.a = 1, .b = 2}, .c = 3} as DerivedClass` would actually be `{.base = {.a = 1, .b = 2}, .c = 3}`. We'll then need to transform conversions from `DerivedClass` to `BaseClass` into accesses of the `.base` member, perhaps in the interpreter's `Convert` function, for both `LValue`s referring to classes and to `NominalClassValue`s.

I ended up having to store the base `NominalClassValue` in a dedicated member of the child class. Storing it in the `init_` struct made things pretty awkward when trying to get and set members. So a bit cleaner, and ticks all the boxes.
I didn't see a way to easily simplify the implicit conversion check, given we need to destination to infer what "base" means. But it still looks acceptable to me.

Pixep commented on explorer/interpreter/type_checker.h:
```diff
@@ -337,6 +337,17 @@ class TypeChecker {
+
+  // Flattens a nested struct used for class initialization that uses the
+  // `.base=` attribute
+  auto FlattenClassInitStruct(SourceLocation source_loc,
```
- [ ] Remove unused declaration

Pixep replied:
Done

Pixep committed:
Explorer: remove unused function declaration

jonmeow commented on explorer/interpreter/value.h:
```diff
@@ -326,19 +326,29 @@ class StructValue : public Value {
+    return base_;
+  }
+
+  static const std::string base_field;
```
Can you switch this to something more like `constexpr llvm::StringLiteral BaseField("base")`? I think we should try to avoid instantiating class types in a static context [per style](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables), and llvm::StringLiteral is intended for this purpose.

Also [per style](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables), static constants should go above the constructor.

Pixep replied:
Good to know, thank you! Done

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -106,6 +107,12 @@ class Interpreter {
                Nonnull<const Value*> destination_type,
                SourceLocation source_loc) -> ErrorOr<Nonnull<const Value*>>;
 
+  // Instantiate a class and its base class(es) from an init struct
```
```suggestion
  // Instantiate a class and its base class(es) from an init struct.
```

Pixep replied:
Done

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -663,6 +671,51 @@ auto Interpreter::InstantiateWitness(Nonnull<const Witness*> witness)
+  std::optional<Nonnull<const NominalClassValue*>> base_instance;
+  for (const auto& field : init_struct->elements()) {
+    if (field.name == NominalClassValue::base_field) {
+      if (!class_type->base().has_value()) {
+        CARBON_FATAL() << "Invalid 'base' field for class '"
+                       << class_type->declaration().name()
+                       << "' without base class.";
+      }
```
```suggestion
      CARBON_CHECK(class_type->base().has_value()) << "Invalid 'base' field for class '"
                       << class_type->declaration().name()
                       << "' without base class.";
```

Pixep replied:
Much better :). Will remember that one. Done.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -436,40 +437,68 @@ auto TypeChecker::FieldTypesImplicitlyConvertible(
+    auto base_fields = FieldTypesWithBase(*class_type.base().value());
+    fields.emplace_back(
+        NamedValue{.name = NominalClassValue::base_field,
+                   .value = (new StructType(std::move(base_fields)))});
```
```suggestion
                   .value = new StructType(std::move(base_fields))});
```

FWIW I'd find this easier to read without the extra parens. Extra parens can also be risky in LLVM, since they're used to silence certain warnings.

Pixep replied:
Agree. Done.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -436,40 +437,68 @@ auto TypeChecker::FieldTypesImplicitlyConvertible(
+                                       impl_scope)) {
+    return false;
+  }
+  if (base_value.has_value()) {
+    if (!IsImplicitlyConvertible(base_value.value(), dest_class.base().value(),
+                                 impl_scope, allow_user_defined_conversions)) {
+      return false;
+    }
+  }
```
```suggestion
  if (base_value.has_value() && !IsImplicitlyConvertible(base_value.value(), dest_class.base().value(),
                                 impl_scope, allow_user_defined_conversions)) {
      return false;
  }
```

I won't comment on the `if` series, but I'll always prefer && instead of nesting.

Pixep replied:
Yeah its not the most elegant code. Its grouped logically, because putting all the conditions together makes it very hard to read or make sense of. But definitely agree for this one.

jonmeow commented on explorer/interpreter/value.cpp:
```diff
@@ -89,7 +107,8 @@ static auto GetMember(Nonnull<Arena*> arena, Nonnull<const Value*> v,
       // Look for a field.
       if (std::optional<Nonnull<const Value*>> field =
-              cast<StructValue>(object.inits()).FindField(f)) {
+              FindClassField(&object, f);
+          field.has_value()) {
```
```suggestion
              FindClassField(&object, f)) {
```

Pixep replied:
Yep... Done.

jonmeow commented on explorer/interpreter/value.cpp:
```diff
@@ -33,6 +38,19 @@ auto StructValue::FindField(std::string_view name) const
+static auto FindClassField(Nonnull<const NominalClassValue*> object,
+                           std::string_view name)
+    -> std::optional<Nonnull<const Value*>> {
+  if (auto field = cast<StructValue>(object->inits()).FindField(name);
+      field.has_value()) {
```
```suggestion
  if (auto field = cast<StructValue>(object->inits()).FindField(name)) {
```
The explicit field.has_value() isn't necessary here since the `if` will handle optional, and I think we generally aren't adding it.

Pixep replied:
Done

jonmeow reviewed (commented):
I think this looks fine and wanted to give some code comments, but I'll see if zygoloid will finish the review.

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
Explorer: fix style, minor improvements

zygoloid commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -106,6 +107,12 @@ class Interpreter {
                Nonnull<const Value*> destination_type,
                SourceLocation source_loc) -> ErrorOr<Nonnull<const Value*>>;
 
+  // Instantiate a class and its base class(es) from an init struct.
+  auto InstantiateClassWithBase(Nonnull<const StructValue*> init,
```
In the interpreter, we use `Instantiate` to mean specifically substituting symbolic names within a value for the corresponding in-scope values of those names, for example "instantiating" the value `Vector(T)` as `Vector(i32)` when `T = i32` is in scope -- somewhat like template instantiation in C++.

This function seems to be doing something unrelated -- more like creating an instance of a class type. Please can you rename this to avoid confusion? Maybe `ConstructClassFromStruct` or `ConvertStructToClass` or something would be clearer? We use the word `Convert` in the interpreter already for this kind of thing.

Pixep replied:
Sure thing, done

zygoloid commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -615,7 +622,8 @@ auto Interpreter::InstantiateType(Nonnull<const Value*> type,
           InstantiateBindings(&class_type.bindings(), source_loc));
-      return arena_->New<NominalClassType>(&class_type.declaration(), bindings);
+      return arena_->New<NominalClassType>(&class_type.declaration(), bindings,
+                                           class_type.base());
```
Is the `base` value here intended to have already been substituted? Eg, given:
```
class A(T:! Type) {}
class B(T:! Type) extends A(T) {}
fn F[T:! Type](b: B(T)) {}
```
... in a call to `F` with `T = i32`, is the type of `B` supposed to have a base of `A(T)` or a base of `A(i32)`? If the latter, then you'll need to call `InstantiateType` here on the base to replace the `T` with `i32` during evaluation.

Pixep replied:
The base should be subsituted as well yes. Thanks!

Pixep replied:
Runnning a basic parametrized base class example fails:
```
base class A(T:! Type) {}
// CHECK:STDERR: COMPILATION ERROR: {{.*}}/explorer/testdata/class/fail_parametrized_base_class.carbon:[[@LINE+1]]: could not resolve 'T'
class B(T:! Type) extends A(T) {}
fn F[T:! Type](b: B(T)) {}
```
Digging a bit, the resolution error occurs while checking an `ImplDeclaration`, specifically `enclosing_scope.Resolve("Self", impl.source_loc()).ok()`. I started looking in the parser to see if the class declaration had an issue, but unsure yet. Still pretty unfamiliar with how generics in general are handled, so I could be missing something.

zygoloid commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -663,6 +671,49 @@ auto Interpreter::InstantiateWitness(Nonnull<const Witness*> witness)
+      CARBON_CHECK(class_type->base().has_value())
+          << "Invalid 'base' field for class '"
+          << class_type->declaration().name() << "' without base class.";
+      switch (field.value->kind()) {
+        case Value::Kind::NominalClassValue: {
+          base_instance = cast<NominalClassValue>(field.value);
+          break;
+        }
+        case Value::Kind::StructValue: {
+          CARBON_ASSIGN_OR_RETURN(
+              auto base,
+              Convert(field.value, class_type->base().value(), source_loc));
+          base_instance = cast<NominalClassValue>(base);
+          break;
+        }
+        default: {
+          CARBON_FATAL()
+              << "Unhandled base value type while initializing class '"
+              << class_type->declaration().name() << "' from a struct";
+        }
+      }
```
It would seem simpler to unconditionally call `Convert` here, as you do in the `StructValue` case:
```suggestion
      CARBON_ASSIGN_OR_RETURN(
          auto base,
          Convert(field.value, class_type->base().value(), source_loc));
      base_instance = cast<NominalClassValue>(base);
```
Would that work? If so, that'd make it easier to add new kinds of built-in conversion to class type -- they'd get picked up here automatically.

Pixep replied:
Good call, it does, much nicer :)

zygoloid commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -663,6 +671,49 @@ auto Interpreter::InstantiateWitness(Nonnull<const Witness*> witness)
+      struct_values.push_back(field);
+    }
+  }
+  // Instantiate the `destination_type` to obtain the runtime
+  // type of the object.
+  CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> inst_class,
+                          InstantiateType(class_type, source_loc));
```
I think this should be done once, up-front, before you start converting, and as described above, `InstantiateType` should itself recursively instantiate the base class. That way, the `NominalClassValue` you create below will have a type with a correct `base` type rather than one that still contains placeholders.

Pixep replied:
Agree, Done

zygoloid commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -436,40 +437,67 @@ auto TypeChecker::FieldTypesImplicitlyConvertible(
+  return fields;
+}
+
+auto TypeChecker::StructImplicitlyConvertibleToClass(
+    const StructType& source_struct, const NominalClassType& dest_class,
+    const ImplScope& impl_scope, bool allow_user_defined_conversions) const
+    -> bool {
+  std::vector<NamedValue> struct_fields{source_struct.fields()};
+  const auto base_it = std::find_if(
+      struct_fields.begin(), struct_fields.end(), [](const auto& field) {
+        return field.name == NominalClassValue::BaseField;
+      });
+  std::optional<const Value*> base_value;
+  if (base_it != struct_fields.end()) {
+    base_value = base_it->value;
+    struct_fields.erase(base_it);
+  }
+  if (base_value.has_value() != dest_class.base().has_value()) {
+    return false;
+  }
+  if (!FieldTypesImplicitlyConvertible(struct_fields, FieldTypes(dest_class),
+                                       impl_scope)) {
+    return false;
+  }
+  if (base_value.has_value() &&
+      !IsImplicitlyConvertible(base_value.value(), dest_class.base().value(),
+                               impl_scope, allow_user_defined_conversions)) {
+    return false;
+  }
+  return true;
+}
```
This seems like it's duplicating the work we do to determine if struct types are implicitly convertible. Perhaps we could call `FieldTypesWithBase(dest_class)` and convert the result to a `StructType` to find the struct type that's the storage type of the class type, then check whether `source_struct` is implicitly convertible to that struct type.

Pixep replied:
Right, and I think I initially missed your point (the same) when you had a similar comment on the first draft. Updated the "ClassFieldsWithBase" to have `base` be a class type, which opened up the door for drastic simplification here. Done.

zygoloid commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -1910,7 +1938,8 @@ auto TypeChecker::SubstituteImpl(const Bindings& bindings,
               &class_type.declaration(),
-              substitute_into_bindings(&class_type.bindings()));
+              substitute_into_bindings(&class_type.bindings()),
+              class_type.base());
```
I expect that we need to recursively substitute into this base class type too: call `SubstituteImpl` on the old base and pass the new value into here.

Pixep replied:
Thanks! Done.

zygoloid commented on explorer/interpreter/value.cpp:
```diff
@@ -33,6 +36,18 @@ auto StructValue::FindField(std::string_view name) const
   return std::nullopt;
 }
 
+static auto FindClassField(Nonnull<const NominalClassValue*> object,
```
Just a note, no change requested:

I think this approach is fine for now, but in the longer term, the type checker should be inserting conversions where necessary to convert derived class expressions to base class expressions so that the interpreter only ever needs to look for a field in the current class, rather than searching its base classes. (Once we support explicitly disambiguating a lookup with `object.(Derived.Field)` vs `object.(Base.Field)`, I think we'll need this.)

Pixep replied:
Agree, I didn't see a way to leverage / test that today, without explicit disambiguation, and kept that aside for now.

zygoloid commented on explorer/testdata/class/base_constructor.carbon:
```diff
@@ -0,0 +1,34 @@
+
+class B extends A {
+  fn Create() -> Self {
+    // Note: This syntax is supported per the design, but currently XFAIL for the explorer.
```
Looks like this is working, is the comment out of date?

Pixep replied:
Yes thank you, done.

zygoloid reviewed (commented):
Thanks, I think this is in good shape. Some comments, mostly around the case where the base class involves a generic parameter (eg, `class Derived(T:! Type) extgends Base(T) {...}`).

Pixep committed:
Explorer: various improvements, generics wip

Pixep committed:
Explorer: refactor implicit conversion logic for classes

Pixep committed:
Explorer: removed unused include

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
Explorer: add clarifying comment to NominalClassType::base_type()

Pixep commented:
Design changes requested have been applied, and notes updated. Two things to note though:
* Generics and specifically parameterized  base classes are not working AFAIS (added that to the notes)
* We have two methods returning a "class type", `NominalClassType::base()` and `ClassDeclaration::base_type()`. The later is needed to resolve parametrized entities (though not working currently), and the former to hold substituted & instantiated base class type. I find it not super clear, nor could I find a quick good name for it, so comment it is. I suppose this is acceptable given I expect this part of the code to change significantly to support generics in a proper way.

Pixep committed:
Explorer: substitue members relative to their class

Fixes an issue causing base class members with generic types to be substituted

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep commented:
With the latest commit, parametrized base classes with their final type set in the class declaration are OK, i.e. `class B extends A(i32)`. Situation where they're not known in advance are not, i.e. `class B(T:! Type) extends A(T)`, as we fail to resolve `T`.

zygoloid commented on explorer/ast/declaration.h:
```diff
@@ -295,12 +293,24 @@ class ClassDeclaration : public Declaration {
+  auto base_expr() const -> std::optional<Nonnull<Expression*>> {
+    return base_expr_;
+  }
   auto base() const -> std::optional<Nonnull<const ClassDeclaration*>> {
     return base_;
   }
   void set_base(Nonnull<const ClassDeclaration*> base_decl) {
     base_ = base_decl;
   }
```
Can we remove this? It doesn't seem like it can be correct to use this when the base can be something like `X(i32)`, and so any callers should be using `base_type` instead. Are there any callers of `base` left?

Pixep replied:
Good point, removed and swapped `base()` calls for `base_type()->declaration()`;

zygoloid commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -663,6 +677,35 @@ auto Interpreter::InstantiateWitness(Nonnull<const Witness*> witness)
+    -> ErrorOr<Nonnull<NominalClassValue*>> {
+  std::vector<NamedValue> struct_values;
+  std::optional<Nonnull<const NominalClassValue*>> base_instance;
+  // Instantiate the `destination_type` to obtain the runtime
+  // type of the object.
+  CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> inst_class,
+                          InstantiateType(class_type, source_loc));
```
Future work (not something I want to see addressed in this PR):

I think this instantiation step should be done higher up the call stack, by the caller of `Convert` -- generally speaking, when the interpreter pulls a type out of the AST (eg, when looking at the `static_type` of an expression), it should immediately instantiate the type to find the proper interpretation of that type in the current context. We shouldn't be passing around uninstantiated types, because that will lead to a lack of clarity over whose responsibility it is to instantiate and to risks that types will either be instantiated repeatedly or that instantiation will be forgotten. (Doing the instantiation in the `Step` function also prepares us for switching to doing non-recursive type instantiation as an `Action`.)

The status quo seems to be that some callers of `Convert` call `InstantiateType`, and others don't, so we sometimes instantiate the type twice (when converting to a class type), sometimes once, and sometimes zero times, depending on what we're converting to, and in which context.

This is a pre-existing problem, not one introduced in this pull request, but this change does seem to be making it somewhat more pressing, given that the call to `Convert` below, when converting the base class, can recursively land back in this function, meaning that in a depth-N inheritance hierarchy we can instantiate N^2 class types.

zygoloid commented on explorer/interpreter/value.cpp:
```diff
@@ -271,6 +299,9 @@ void Value::Print(llvm::raw_ostream& out) const {
       const auto& s = cast<NominalClassValue>(*this);
       out << cast<NominalClassType>(s.type()).declaration().name() << s.inits();
+      if (s.base().has_value()) {
+        out << ", " << *s.base().value();
```
I think a different sigil instead of `,` here would be better. For example, right now we'll print things like:
```
Derived{.derived_field = 5}, Base{.base_field = 7}
```
... which looks OK by itself but will be confusing if we're printing a tuple value. For example:
```
(1, Derived{.derived_field = 5}, Base{.base_field = 7}, 3)
```
... looks like a 4-tuple instead of a 3-tuple. Maybe:
```suggestion
        out << " base " << *s.base().value();
```
... so we get:
```
(1, Derived{.derived_field = 5} base Base{.base_field = 7}, 3)
```

(Much as it might be nice to produce valid Carbon here, `{.derived_field = 5, .base = {.base_field = 7} as Base} as Derived` seems to be less readable as well as being more awkward to print, so I'm not going to ask for that.)

zygoloid reviewed (commented):
Looks good, only a couple of actionable comments.

Pixep committed:
Explorer: disambiguate base class output

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

Pixep committed:
Explorer: remove unecessary ClassDeclaration::base()

Pixep commented:
> Looks good, only a couple of actionable comments.

Addressed actionable comments ✅, thanks!
